File: JavaSimpleBackend/src/main/java/com/atesti/service/RadniNaloziService.java
package com.atesti.service;

import com.atesti.dto.RadniNalogRequest;
import com.atesti.dto.UskoroIsticeResponse;
import com.atesti.entity.*;
import com.atesti.exception.BadRequestException;
import com.atesti.exception.ResourceNotFoundException;
import com.atesti.repository.*;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class RadniNaloziService {

    private final RadniNalogRepository radniNalogRepository;
    private final NaruciteljRepository naruciteljRepository;
    private final AktivnostRepository aktivnostRepository;
    private final UskoroIsticeRepository uskoroIsticeRepository;
    private final UserRepository userRepository;
    private final ObjectMapper objectMapper;

    public String getNextBrojNaloga() {
        long count = radniNalogRepository.count();
        long sequenceNumber = count + 1;

        while (true) {
            String candidate = String.format("RN%03d", sequenceNumber);
            Optional<RadniNalog> existing = radniNalogRepository.findByBrojNaloga(candidate);
            if (existing.isEmpty()) {
                return candidate;
            }
            sequenceNumber++;
        }
    }

    public List<UskoroIsticeResponse> getUskoroIstice(int days) {
        LocalDate today = LocalDate.now();
        LocalDate threshold = today.plusDays(days);

        List<UskoroIstice> items = uskoroIsticeRepository
                .findByIsActiveTrueAndDatumIstekaBetweenOrderByDatumIstekaAsc(today, threshold);

        return items.stream().map(item -> UskoroIsticeResponse.builder()
                .id(item.getId())
                .narucitelj(item.getNarucitelj() != null ? item.getNarucitelj().getName() : null)
                .radniNalog(item.getRadniNalog() != null ? item.getRadniNalog().getBrojNaloga() : null)
                .aktivnost(item.getAktivnost())
                .datumIsteka(item.getDatumIsteka())
                .isActive(item.getIsActive())
                .build()
        ).collect(Collectors.toList());
    }

    public List<RadniNalog> getAllRadniNalozi() {
        return radniNalogRepository.findAll();
    }

    public RadniNalog getRadniNalogById(Long id) {
        return radniNalogRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Radni nalog not found"));
    }

    public List<RadniNalog> getRadniNaloziByNarucitelj(Long naruciteljId) {
        return radniNalogRepository.findByNaruciteljId(naruciteljId);
    }

    @Transactional
    public RadniNalog createRadniNalog(RadniNalogRequest request) {
        if (request.getAktivnosti() == null) {
            throw new BadRequestException("Error creating radni nalog: Aktivnosti field is required and must be an array");
        }

        List<Long> aktivnostiList = normalizeAktivnostiInput(request.getAktivnosti());

        Long naruciteljId = request.getNaruciteljId() != null ? request.getNaruciteljId() : request.getNaruciteljIdAlt();
        Narucitelj narucitelj = naruciteljRepository.findById(naruciteljId)
                .orElseThrow(() -> new BadRequestException("Error creating radni nalog: Narucitelj not found"));

        String brojNaloga = request.getBrojNaloga();
        if (brojNaloga == null || brojNaloga.isBlank()) {
            brojNaloga = getNextBrojNaloga();
        }

        if (!brojNaloga.matches("^RN\\d{3}$")) {
            throw new BadRequestException("Error creating radni nalog: Broj naloga must be in format RN followed by 3 digits, e.g. RN001");
        }

        validateAktivnosti(aktivnostiList);

        RadniNalog nalog = RadniNalog.builder()
                .brojNaloga(brojNaloga)
                .narucitelj(narucitelj)
                .datum(parseDatum(request.getDatum()))
                .objekt(request.getObjekt())
                .fakturirano(request.getFakturirano() != null ? request.getFakturirano() : false)
                .zavrseno(request.getZavrseno() != null ? request.getZavrseno() : false)
                .opis(request.getOpis())
                .brojPonude(request.getBrojPonude())
                .brojRacuna(request.getBrojRacuna())
                .narudzbenica(request.getNarudzbenica())
                .ugovor(request.getUgovor())
                .aktivnosti(serializeAktivnosti(aktivnostiList))
                .pdfUrl(request.getPdfUrl())
                .build();

        nalog = radniNalogRepository.save(nalog);
        replaceUskoroIsticeEntries(nalog, aktivnostiList);

        return nalog;
    }

    @Transactional
    public RadniNalog updateRadniNalog(Long id, RadniNalogRequest request) {
        RadniNalog nalog = radniNalogRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Error updating radni nalog: Radni nalog not found"));

        if (request.getAktivnosti() == null) {
            throw new BadRequestException("Error updating radni nalog: Aktivnosti field is required and must be an array");
        }

        List<Long> aktivnostiList = normalizeAktivnostiInput(request.getAktivnosti());

        Long nextNaruciteljId = request.getNaruciteljIdAlt() != null ? request.getNaruciteljIdAlt()
                : (request.getNaruciteljId() != null ? request.getNaruciteljId() : nalog.getNaruciteljId());

        if (nextNaruciteljId != null) {
            Narucitelj narucitelj = naruciteljRepository.findById(nextNaruciteljId)
                    .orElseThrow(() -> new BadRequestException("Error updating radni nalog: Narucitelj not found"));
            nalog.setNarucitelj(narucitelj);
        }

        String brojNaloga = request.getBrojNaloga() != null ? request.getBrojNaloga() : nalog.getBrojNaloga();
        if (!brojNaloga.matches("^RN\\d{3}$")) {
            throw new BadRequestException("Error updating radni nalog: Broj naloga must be in format RN followed by 3 digits, e.g. RN001");
        }
        nalog.setBrojNaloga(brojNaloga);

        validateAktivnosti(aktivnostiList);

        if (request.getDatum() != null) nalog.setDatum(parseDatum(request.getDatum()));
        if (request.getObjekt() != null) nalog.setObjekt(request.getObjekt());
        if (request.getFakturirano() != null) nalog.setFakturirano(request.getFakturirano());
        if (request.getZavrseno() != null) nalog.setZavrseno(request.getZavrseno());
        if (request.getOpis() != null) nalog.setOpis(request.getOpis());
        if (request.getBrojPonude() != null) nalog.setBrojPonude(request.getBrojPonude());
        if (request.getBrojRacuna() != null) nalog.setBrojRacuna(request.getBrojRacuna());
        if (request.getNarudzbenica() != null) nalog.setNarudzbenica(request.getNarudzbenica());
        if (request.getUgovor() != null) nalog.setUgovor(request.getUgovor());
        if (request.getPdfUrl() != null) nalog.setPdfUrl(request.getPdfUrl());

        nalog.setAktivnosti(serializeAktivnosti(aktivnostiList));
        nalog = radniNalogRepository.save(nalog);

        replaceUskoroIsticeEntries(nalog, aktivnostiList);

        return nalog;
    }

    @Transactional
    public RadniNalog deleteRadniNalog(Long id) {
        RadniNalog nalog = radniNalogRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Error deleting radni nalog: Radni nalog not found"));

        if (!nalog.getBrojNaloga().matches("^RN\\d{3}$")) {
            throw new BadRequestException("Error deleting radni nalog: Broj naloga must be in format RN followed by 3 digits, e.g. RN001");
        }

        radniNalogRepository.delete(nalog);
        return nalog;
    }

    @Transactional
    public RadniNalog assignWorker(Long id, Long assignedUserId) {
        RadniNalog nalog = radniNalogRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Radni nalog not found"));

        if (assignedUserId == null) {
            nalog.setAssignedUser(null);
        } else {
            User user = userRepository.findById(assignedUserId)
                    .orElseThrow(() -> new ResourceNotFoundException("User not found"));
            nalog.setAssignedUser(user);
        }

        return radniNalogRepository.save(nalog);
    }

    private List<Long> normalizeAktivnostiInput(List<Object> aktivnostiInput) {
        if (aktivnostiInput == null) {
            return new ArrayList<>();
        }

        List<Long> sanitized = new ArrayList<>();
        for (Object item : aktivnostiInput) {
            Long parsedId = null;

            if (item instanceof Number) {
                long val = ((Number) item).longValue();
                if (val > 0) parsedId = val;
            } else if (item instanceof String) {
                String trimmed = ((String) item).trim();
                if (trimmed.matches("^\\d+$")) {
                    long val = Long.parseLong(trimmed);
                    if (val > 0) parsedId = val;
                }
            }

            if (parsedId == null) {
                throw new BadRequestException("Aktivnosti must be an array of numeric IDs");
            }

            if (!sanitized.contains(parsedId)) {
                sanitized.add(parsedId);
            }
        }

        return sanitized;
    }

    private void validateAktivnosti(List<Long> aktivnostiList) {
        if (aktivnostiList.isEmpty()) return;

        List<Aktivnost> found = aktivnostRepository.findByIdInAndIsActiveTrue(aktivnostiList);
        Set<Long> foundIds = found.stream().map(Aktivnost::getId).collect(Collectors.toSet());

        List<Long> missingIds = aktivnostiList.stream()
                .filter(id -> !foundIds.contains(id))
                .collect(Collectors.toList());

        if (!missingIds.isEmpty()) {
            throw new BadRequestException("Aktivnosti IDs are invalid or inactive: " +
                    missingIds.stream().map(String::valueOf).collect(Collectors.joining(", ")));
        }
    }

    private void replaceUskoroIsticeEntries(RadniNalog nalog, List<Long> aktivnostiList) {
        uskoroIsticeRepository.deleteByRadniNalogId(nalog.getId());

        if (aktivnostiList.isEmpty()) return;

        List<Aktivnost> aktivnostiConfig = aktivnostRepository.findByIdInAndIsActiveTrue(aktivnostiList);
        Map<Long, Aktivnost> configById = aktivnostiConfig.stream()
                .collect(Collectors.toMap(Aktivnost::getId, a -> a));

        for (Long aktivnostId : aktivnostiList) {
            Aktivnost config = configById.get(aktivnostId);
            if (config == null) continue;

            int rokTrajanjaMjeseci = config.getRokTrajanja();
            if (rokTrajanjaMjeseci <= 0) continue;

            LocalDate datumIsteka = nalog.getDatum().toLocalDate().plusMonths(rokTrajanjaMjeseci);

            UskoroIstice entry = UskoroIstice.builder()
                    .naruciteljId(nalog.getNaruciteljId())
                    .radniNalogId(nalog.getId())
                    .aktivnost(config.getAktivnost())
                    .datumIsteka(datumIsteka)
                    .isActive(true)
                    .build();

            uskoroIsticeRepository.save(entry);
        }
    }

    private String serializeAktivnosti(List<Long> aktivnostiList) {
        try {
            return objectMapper.writeValueAsString(aktivnostiList);
        } catch (JsonProcessingException e) {
            return "[]";
        }
    }

    private LocalDateTime parseDatum(String datum) {
        if (datum == null) return LocalDateTime.now();
        try {
            return LocalDateTime.parse(datum);
        } catch (Exception e) {
            try {
                return LocalDate.parse(datum).atStartOfDay();
            } catch (Exception e2) {
                try {
                    return LocalDateTime.parse(datum, DateTimeFormatter.ISO_DATE_TIME);
                } catch (Exception e3) {
                    return LocalDateTime.now();
                }
            }
        }
    }
}

File: backend/services/radniNaloziService.js
const db = require('../models');

class RadniNaloziService {
  constructor() {
    this._aktivnostiSupportsArrayCache = null;
  }

  formatDateOnly(dateValue) {
    const date = new Date(dateValue);
    const year = date.getFullYear();
    const month = String(date.getMonth() + 1).padStart(2, '0');
    const day = String(date.getDate()).padStart(2, '0');
    return `${year}-${month}-${day}`;
  }

  calculateDatumIsteka(datum, trajanjeMjeseci) {
    const datumIstekaDate = new Date(datum);
    datumIstekaDate.setMonth(datumIstekaDate.getMonth() + trajanjeMjeseci);
    return datumIstekaDate;
  }

  normalizeAktivnostiInput(aktivnostiInput) {
    if (aktivnostiInput === null || aktivnostiInput === undefined) {
      return [];
    }

    if (!Array.isArray(aktivnostiInput)) {
      throw new Error('Aktivnosti must be an array');
    }

    const sanitized = [];

    for (const item of aktivnostiInput) {
      let parsedId = null;

      if (typeof item === 'number' && Number.isInteger(item) && item > 0) {
        parsedId = item;
      } else if (typeof item === 'string') {
        const trimmed = item.trim();
        if (/^\d+$/.test(trimmed)) {
          const numericValue = Number(trimmed);
          if (Number.isInteger(numericValue) && numericValue > 0) {
            parsedId = numericValue;
          }
        }
      }

      if (!parsedId) {
        throw new Error('Aktivnosti must be an array of numeric IDs');
      }

      if (!sanitized.includes(parsedId)) {
        sanitized.push(parsedId);
      }
    }

    return sanitized;
  }

  async supportsArrayAktivnostiStorage() {
    if (this._aktivnostiSupportsArrayCache !== null) {
      return this._aktivnostiSupportsArrayCache;
    }

    const tableDefinition = await db.sequelize.getQueryInterface().describeTable('radni_nalozi');
    const aktivnostiColumn = tableDefinition.aktivnosti;
    const columnType = aktivnostiColumn && typeof aktivnostiColumn.type === 'string'
      ? aktivnostiColumn.type.toUpperCase()
      : '';

    this._aktivnostiSupportsArrayCache = columnType.includes('TEXT');
    return this._aktivnostiSupportsArrayCache;
  }

  serializeAktivnostiForStorage(aktivnostiList, supportsArrayStorage) {
    if (supportsArrayStorage) {
      return JSON.stringify(aktivnostiList);
    }

    return aktivnostiList.length > 0 ? aktivnostiList[0] : null;
  }

  async replaceUskoroIsticeEntries(nalog, aktivnostiList, transaction) {
    await db.uskoro_istice.destroy({
      where: { radni_nalog_id: nalog.id },
      transaction
    });

    if (aktivnostiList.length === 0) {
      return;
    }

    const aktivnostiConfig = await db.aktivnost.findAll({
      where: {
        id: {
          [db.Sequelize.Op.in]: aktivnostiList
        },
        isActive: true
      },
      transaction
    });

    const configById = aktivnostiConfig.reduce((accumulator, item) => {
      accumulator[item.id] = {
        aktivnost: item.aktivnost,
        rokTrajanja: item.rokTrajanja
      };
      return accumulator;
    }, {});

    const missingIds = aktivnostiList.filter((id) => !configById[id]);
    if (missingIds.length > 0) {
      throw new Error(`Aktivnosti IDs are invalid or inactive: ${missingIds.join(', ')}`);
    }

    for (const aktivnostId of aktivnostiList) {
      const config = configById[aktivnostId];
      const rokTrajanjaMjeseci = config.rokTrajanja;

      const datumIsteka = this.calculateDatumIsteka(nalog.datum, rokTrajanjaMjeseci);

      if(rokTrajanjaMjeseci <= 0){
        continue; // skip creating uskoro istice entry for non-positive durations
      }
      await db.uskoro_istice.create({
        narucitelj_id: nalog.narucitelj_id,
        radni_nalog_id: nalog.id,
        aktivnost: config.aktivnost,
        datumIsteka: this.formatDateOnly(datumIsteka),
        isActive: true
      }, { transaction });
    }
  }

  async getUskoroIstice(days = 30) {
    try {
      const now = new Date();
      const today = new Date(now.getFullYear(), now.getMonth(), now.getDate());
      const threshold = new Date(today);
      threshold.setDate(threshold.getDate() + days);

      const items = await db.uskoro_istice.findAll({
        where: {
          isActive: true,
          datumIsteka: {
            [db.Sequelize.Op.between]: [this.formatDateOnly(today), this.formatDateOnly(threshold)]
          }
        },
        include: [
          {
            model: db.narucitelj,
            as: 'narucitelj'
          },
          {
            model: db.radni_nalog,
            as: 'radniNalog'
          }
        ],
        order: [['datumIsteka', 'ASC']]
      });

      const results = items.map((item) => ({
        id: item.id,
        narucitelj: item.narucitelj ? item.narucitelj.name : null,
        radniNalog: item.radniNalog ? item.radniNalog.brojNaloga : null,
        aktivnost: item.aktivnost,
        datumIsteka: item.datumIsteka,
        isActive: item.isActive
      }));

      return results;
    } catch (error) {
      throw new Error(`Error fetching uskoro istice: ${error.message}`);
    }
  }

  async assignWorker(nalogId, userId) {
    const nalog = await db.radni_nalog.findByPk(nalogId);
    if (!nalog) throw new Error('Radni nalog not found');

    if (userId) {
      const user = await db.user.findByPk(userId);
      if (!user) throw new Error('User not found');
    }

    nalog.assignedUserId = userId || null;
    await nalog.save();
    return nalog;
  }

  async getAllRadniNalozi() {
    try {
      return await db.radni_nalog.findAll({
        include: [
          {
            model: db.narucitelj,
            as: 'narucitelj'
          },
          {
            model: db.document,
            as: 'documents',
            required: false
          },
          {
            model: db.note,
            as: 'notes',
            required: false
          },
          {
            model: db.user,
            as: 'assignedUser',
            attributes: ['id', 'username', 'ime', 'prezime'],
            required: false
          }
        ]
      });
    } catch (error) {
      throw new Error(`Error fetching radni nalozi: ${error.message}`);
    }
  }

  async getRadniNalogById(id) {
    try {
      const nalog = await db.radni_nalog.findByPk(id, {
        include: [
          {
            model: db.narucitelj,
            as: 'narucitelj'
          },
          {
            model: db.document,
            as: 'documents',
            required: false
          },
          {
            model: db.note,
            as: 'notes',
            required: false
          },
          {
            model: db.user,
            as: 'assignedUser',
            attributes: ['id', 'username', 'ime', 'prezime'],
            required: false
          }
        ]
      });

      if (!nalog) {
        throw new Error('Radni nalog not found');
      }

      return nalog;
    } catch (error) {
      throw new Error(`Error fetching radni nalog: ${error.message}`);
    }
  }

  async getRadniNaloziByNarucitelj(naruciteljiId) {
    try {
      const nalozi = await db.radni_nalog.findAll({
        where: {
          narucitelj_id: naruciteljiId
        },
        include: [
          {
            model: db.narucitelj,
            as: 'narucitelj'
          },
          {
            model: db.document,
            as: 'documents',
            required: false
          },
          {
            model: db.note,
            as: 'notes',
            required: false
          }
        ]
      });

      return nalozi;
    } catch (error) {
      throw new Error(`Error fetching radni nalozi: ${error.message}`);
    }
  }

  async createRadniNalog(data) {
    try {
      if (!Object.prototype.hasOwnProperty.call(data, 'aktivnosti')) {
        throw new Error('Aktivnosti field is required and must be an array');
      }

      const aktivnostiList = this.normalizeAktivnostiInput(data.aktivnosti);
      const supportsArrayStorage = await this.supportsArrayAktivnostiStorage();

      if (!supportsArrayStorage && aktivnostiList.length > 1) {
        throw new Error('Database schema supports only one aktivnost per radni nalog. Run latest migrations to enable multiple aktivnosti.');
      }

      // ensure narucitelj exists before creating
      const klijent = await db.narucitelj.findByPk(data.naruciteljId);
      if (!klijent) {
        throw new Error('Narucitelj not found');
      }
      var nextNumber = data.brojNaloga || null;
      if(nextNumber == null){
        nextNumber = await this.getNextBrojNaloga();
      }
      // Enforce RN + 3 digit format
      if (!/^RN\d{3}$/.test(nextNumber)) {
        throw new Error('Broj naloga must be in format RN followed by 3 digits, e.g. RN001');
      }

      return await db.sequelize.transaction(async (transaction) => {
        const nalog = await db.radni_nalog.create({
          brojNaloga: nextNumber,
          narucitelj_id: data.naruciteljId,
          datum: new Date(data.datum),
          objekt: data.objekt,
          fakturirano: data.fakturirano || false,
          zavrseno: data.zavrseno || false,
          opis: data.opis || null,
          brojPonude: data.brojPonude || null,
          brojRacuna: data.brojRacuna || null,
          narudzbenica: data.narudzbenica || null,
          ugovor: data.ugovor || null,
          aktivnosti: this.serializeAktivnostiForStorage(aktivnostiList, supportsArrayStorage),
          pdfUrl: data.pdfUrl || null
        }, { transaction });

        await this.replaceUskoroIsticeEntries(nalog, aktivnostiList, transaction);
        return nalog;
      });
    } catch (error) {
      throw new Error(`Error creating radni nalog: ${error.message}`);
    }
  }

  async updateRadniNalog(id, data) {
    try {
      const nalog = await db.radni_nalog.findByPk(id);

      if (!nalog) {
        throw new Error('Radni nalog not found');
      }

      if (!Object.prototype.hasOwnProperty.call(data, 'aktivnosti')) {
        throw new Error('Aktivnosti field is required and must be an array');
      }

      const nextNaruciteljId = data.narucitelj_id !== undefined
        ? data.narucitelj_id
        : (data.naruciteljId !== undefined ? data.naruciteljId : nalog.narucitelj_id);

      const nextAktivnosti = this.normalizeAktivnostiInput(data.aktivnosti);
      const supportsArrayStorage = await this.supportsArrayAktivnostiStorage();

      if (!supportsArrayStorage && nextAktivnosti.length > 1) {
        throw new Error('Database schema supports only one aktivnost per radni nalog. Run latest migrations to enable multiple aktivnosti.');
      }

      // Enforce RN + 3 digit format if brojNaloga is being updated
      let brojNalogaToCheck = data.brojNaloga !== undefined ? data.brojNaloga : nalog.brojNaloga;
      if (!/^RN\d{3}$/.test(brojNalogaToCheck)) {
        throw new Error('Broj naloga must be in format RN followed by 3 digits, e.g. RN001');
      }

      if (nextNaruciteljId !== undefined) {
        const klijent = await db.narucitelj.findByPk(nextNaruciteljId);
        if (!klijent) {
          throw new Error('Narucitelj not found');
        }
      }

      return await db.sequelize.transaction(async (transaction) => {
        const updatedNalog = await nalog.update({
          narucitelj_id: nextNaruciteljId,
          datum: data.datum ? new Date(data.datum) : nalog.datum,
          objekt: data.objekt || nalog.objekt,
          fakturirano: data.fakturirano !== undefined ? data.fakturirano : nalog.fakturirano,
          zavrseno: data.zavrseno !== undefined ? data.zavrseno : nalog.zavrseno,
          opis: data.opis !== undefined ? data.opis : nalog.opis,
          brojPonude: data.brojPonude !== undefined ? data.brojPonude : nalog.brojPonude,
          brojRacuna: data.brojRacuna !== undefined ? data.brojRacuna : nalog.brojRacuna,
          narudzbenica: data.narudzbenica !== undefined ? data.narudzbenica : nalog.narudzbenica,
          ugovor: data.ugovor !== undefined ? data.ugovor : nalog.ugovor,
          aktivnosti: this.serializeAktivnostiForStorage(nextAktivnosti, supportsArrayStorage),
          pdfUrl: data.pdfUrl !== undefined ? data.pdfUrl : nalog.pdfUrl,
          brojNaloga: brojNalogaToCheck
        }, { transaction });

        await this.replaceUskoroIsticeEntries(updatedNalog, nextAktivnosti, transaction);

        return updatedNalog;
      });
    } catch (error) {
      throw new Error(`Error updating radni nalog: ${error.message}`);
    }
  }

  async deleteRadniNalog(id) {
    try {
      const nalog = await db.radni_nalog.findByPk(id);

      if (!nalog) {
        throw new Error('Radni nalog not found');
      }
      // Validate brojNaloga format before delete
      if (!/^RN\d{3}$/.test(nalog.brojNaloga)) {
        throw new Error('Broj naloga must be in format RN followed by 3 digits, e.g. RN001');
      }

      await nalog.destroy();
      return nalog;
    } catch (error) {
      throw new Error(`Error deleting radni nalog: ${error.message}`);
    }
  }

  async getNextBrojNaloga() {
    try {
      let sequenceNumber = (await db.radni_nalog.count()) + 1;

      while (true) {
        const candidate = `RN${String(sequenceNumber).padStart(3, '0')}`;
        const existing = await db.radni_nalog.findOne({
          where: { brojNaloga: candidate },
          attributes: ['id']
        });

        if (!existing) {
          return candidate;
        }

        sequenceNumber += 1;
      }
    } catch (error) {
      throw new Error(`Error generating broj naloga: ${error.message}`);
    }
  }
}

module.exports = new RadniNaloziService();
